docs/DifferentiableFunctions.md

```diff
@@ -0,0 +1,454 @@
+let dense = DenseLayer(weight: [[1, 1], [1, 1]], bias: [1, 1])
+let input: Tensor<Float> = [[3, 3]]
+let seed: Tensor<Float> = [[1, 1]]
+
+let (output, backprop) = dense.appliedForBackpropagation(to: input)
+let (𝛁dense, 𝛁input) = backprop(seed)
+
+dump(𝛁dense)
+// ▿ DenseLayer.AllDifferentiableVariables
+//   - weight: [[3.0, 3.0], [3.0, 3.0]]
+//   - bias: [1.0, 1.0]
+print(𝛁input)
+// [[2.0, 2.0]]
+```
+
+## Providing a custom derivative
+
+Use the `@differentiating` attribute to mark a function as a custom derivative
+for another function. This is useful for registering derivatives for primitive
+operations.
+
+Note: currently, the `@differentiating` attribute can only be used to define
+derivatives for functions in the same module. We plan to lift this limitation
+soon so that derivatives can be retroactively declared for functions in other
+modules - [see this forum
+discussion](https://forums.swift.org/t/help-needed-with-retroactive-differentiability/19927)
+for more information.
+
+```swift
+import Darwin
+
+func sillyExp(_ x: Float) -> Float {
+    let 𝑒 = Float(M_E)
+    print("Taking 𝑒(\(𝑒)) to the power of \(x)!")
+    return pow(𝑒, x)
+}
+
+@differentiating(sillyExp)
+func sillyDerivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
+    let y = sillyExp(x)
+    return (value: y, pullback: { v in v * y })
+}
+
+print(gradient(of: sillyExp)(3))
+// Taking 𝑒(2.7182817) to the power of 3.0!
+// 20.085535
+```
+
+## Constructing a `@differentiable` function from a derivative
+
+Given a function and its derivative, it is possible to construct a
+`@differentiable` version of the function using the
+[`differentiableFunction(from:)`] helper function defined in the standard
+library.
+
+Here's an example:
+```swift
+let multiply: @differentiable (Float, Float) -> Float =
+    differentiableFunction(from: { x, y in (value: x * y, pullback: { v in (v * y, v * x) }) })
+```
+
+Internally, `differentiableFunction(from:)` is defined just using the
+`@differentiating` attribute - there's no extra magic:
+
+```swift
+/// Returns a differentiable function given its derivative.
+public func differentiableFunction<T: Differentiable, R: Differentiable>(
+    from vjp: @escaping (T) -> (value: R, pullback: (R.CotangentVector) -> T.CotangentVector)
+) -> @differentiable (T) -> R {
+    func original(_ x: T) -> R {
+        return vjp(x).value
+    }
+    @differentiating(original)
+    func derivative(_ x: T) -> (value: R, pullback: (R.CotangentVector) -> T.CotangentVector) {
+        return vjp(x)
+    }
+    return original
+}
+```
+
+## `@differentiable` functions and automatic differentiation
+
+Automatic differentiation is the technique used by the compiler to automatically
+compute function derivatives. This document does not go into detail about
+automatic differentiation - but with an understanding of `@differentiable`
+functions and differentiation APIs, one can get a glimpse of how automatic
+differentiation works.
```

Comment:
Perhaps add some lines to imply to our users that "automatic differentiation" is just an implementation detail in the compiler that enables derivative calculation, and that only differentiation APIs and `@differentiable` attributes are what's to be internalized.

Reply:
Adding:
```markdown
*Note: automatic differentiation is an implementation detail in the compiler
that enables differentiation. `@differentiable` functions and differentiation
APIs are the important, visible features to understand.*
```